src/dust/dust.cpp

```diff
@@ -320,6 +457,344 @@ TaskStatus FluxSource(MeshData<Real> *md, const Real dt) {
+
+          const Real dens_g = vmesh(b, gas::prim::density(0), k, j, i) * den0;
+          Real dt_sync = dt * time0;
+          const Real time1 = time * time0;
+
+          geometry::Coords<GEOM> coords(vmesh.GetCoordinates(b), k, j, i);
+          const auto &hx = coords.GetScaleFactors();
+          Real rad = hx[2];
+          if constexpr (GEOM == Coordinates::cylindrical) {
+            rad = hx[1];
+          }
+          const Real Omega_k = 1.0 / std::sqrt(rad) / rad; // code unit
+          // const Real vol1 = coords.Volume() * vol0;
+
+          int nCall1 = 0;
+
+          const Real pres = vmesh(b, gas::prim::pressure(0), k, j, i);
+          const Real cs = std::sqrt(gamma * pres / dens_g * den0) * vel0;
+          const Real omega1 = Omega_k / time0;
+          const int nm = nspecies;
+
+          ScratchPad1D<Real> rhod(mbr.team_scratch(scr_level), nspecies);
+          ScratchPad1D<Real> stime(mbr.team_scratch(scr_level), nspecies);
+          ScratchPad1D<Real> vel(mbr.team_scratch(scr_level), nvel * nspecies);
+          ScratchPad1D<Real> source(mbr.team_scratch(scr_level), nspecies);
+          ScratchPad1D<Real> Q(mbr.team_scratch(scr_level), nspecies);
+          ScratchPad1D<Real> nQs(mbr.team_scratch(scr_level), nspecies);
+          [[maybe_unused]] ScratchPad1D<Real> Q2;
+          if (coag.integrator == 3 && coag.mom_coag) {
+            Q2 = ScratchPad1D<Real>(mbr.team_scratch(scr_level), nspecies);
+          }
+
+          parthenon::par_for_inner(
+              DEFAULT_INNER_LOOP_PATTERN, mbr, 0, nm - 1, [&](const int n) {
+                // for (int n = 0; n < nm; ++n) {
+                stime(n) = vmesh(b, dust::stopping_time(n), k, j, i) * time0;
+                if (vmesh(b, dust::prim::density(n), k, j, i) > dfloor) {
+                  rhod(n) = vmesh(b, dust::prim::density(n), k, j, i) * den0;
+                  vel(0 + n * 3) =
+                      vmesh(b, dust::prim::velocity(VI(n, 0)), k, j, i) * vel0;
+                  vel(1 + n * 3) =
+                      vmesh(b, dust::prim::velocity(VI(n, 1)), k, j, i) * vel0;
+                  vel(2 + n * 3) =
+                      vmesh(b, dust::prim::velocity(VI(n, 2)), k, j, i) * vel0;
+                } else {
+                  rhod(n) = 0.0;
+                  vel(0 + n * 3) = 0.0;
+                  vel(1 + n * 3) = 0.0;
+                  vel(2 + n * 3) = 0.0;
+                }
+              });
+
+          Coagulation::CoagulationOneCell(mbr, i, time1, dt_sync, dens_g, rhod, stime,
+                                          vel, nvel, Q, nQs, alpha, cs, omega1, coag,
+                                          source, nCall1, Q2);
+
+          nCalls(k, j, i) = nCall1;
+
+          // update dust density and velocity after coagulation
+          parthenon::par_for_inner(
+              DEFAULT_INNER_LOOP_PATTERN, mbr, 0, nm - 1, [&](const int n) {
+                // for (int n = 0; n < nspecies; ++n) {
```

Comment:
this is not written with sparsity in mind.

Reply:
within coagulation calling routine, all species info are used.

Reply:
When sparsity is active, we will only want to process coagulation between the active dust bins in a given meshblock (since the dust density will be assumed to be zero for all inactive sparse indices).

That being said, let's defer sparse considerations to a subsequent PR.